### PR TITLE
build(deps): bump luajit to HEAD - 03c31124c

### PR DESCRIPTION
**Pull request by clason**
switch to rolling releases; let's see how our build system likes that...

### CODE REVIEW

**Comment by clason** [comment]
*Narrator: It didn't.*

@dundar 🥺

**Comment by clason** [comment]
Looks like the LuaJIT build scripts can't find or use `git` to get the "rolling release tag", so the binary (used for docgen) is improperly named (`luajit-2.1.`, missing "version").

File: cmake.deps/cmake/BuildLuajit.cmake [hunk 1]
```diff
@@ -30,7 +30,7 @@ function(BuildLuajit)
   if(UNIX)
     add_custom_command(
       TARGET ${_luajit_TARGET}
-      COMMAND ${CMAKE_COMMAND} -E create_symlink luajit-2.1.0-beta3 ${DEPS_BIN_DIR}/${_luajit_TARGET})
+      COMMAND ${CMAKE_COMMAND} -E create_symlink luajit-2.1 ${DEPS_BIN_DIR}/${_luajit_TARGET})
```
Comment (dundargoc) [inline cmake.deps/cmake/BuildLuajit.cmake:33]: This makes the build work for me locally, although it's ugly. I think the dot comes from a luajit bug in the build process, so might be worth waiting a bit to confirm this so we can fix this upstream.

```suggestion
      COMMAND ${CMAKE_COMMAND} -E create_symlink luajit-2.1. ${DEPS_BIN_DIR}/${_luajit_TARGET})
```

**Comment by dundargoc** [comment]
I commited my suggestion mostly because I wanted to see the CI results, feel free to remove afterwards.

**Comment by dundargoc** [comment]
Luajit can't calculate the version since we're currently downloading the tarball which doesn't have the `.git` directory. I see the following options we can take:

1. accept the PR as is.
2. Use `GIT_REPOSITORY` instead of `URL` to specify the git repo as the dependency rather than the tarball. A disadvantage is that we won't be able to use SHA256 to verify if we got the expected code. Another is that if we go this route then the executable name will change with every version change, but this is a minor annoyance that should be easily fixable.

I'd lean for option 1 currently unless we find more advantages of option 2. An advantage might be that we can start specifying the minimum luajit version, assuming the version numbers are monotonically increasing and not random nonsense.

**Comment by clason** [comment]
So I looked a bit into the vanilla build scripts:

1. `make` builds a `luajit` binary under `src/`
2. `make install` copies that to `luajit-2.1.<hash>`
3. as you observed, `<hash>` is empty if the `.git` directory is missing
4. even with a `.git` directory, the `make install` step fails for me with 
```
==== Installing LuaJIT 2.1.1692616192 to . ====
mkdir -p ./bin ./lib ./include/luajit-2.1 ./share/man/man1 ./lib/pkgconfig ./share/luajit-2.1/jit ./share/lua/5.1 ./lib/lua/5.1
cd src && install -m 0755 luajit ./bin/luajit-2.1.1692616192
install: ./bin/luajit-2.1.1692616192: No such file or directory
make: *** [install] Error 71
```

So it looks like the new "rolling release" model still has some rough spots and 
>  assuming the version numbers are monotonically increasing and not random nonsense.

is unfortunately not true.

The most robust option (given the current upstream Makefile) would be to _not_ rely on `make install` and instead manually copy the files ourselves from the source tree instead -- if that is possible.

**Comment by MikePall** [comment]
The last number in the semver is a monotonic *timestamp*, not a hash.

By trusting a git hash you're automatically creating a trust chain down to the individual files. That is, if you trust your git client. Right now, you're trusting the hash of a derivative made by a third party.

The missing symlink was an oversight and is now fixed, thanks to the report by @clason.

`DESTDIR` must be an absolute path, use `$PWD` and not a dot. See https://www.gnu.org/prep/standards/html_node/DESTDIR.html.

And, yes, please cache all repos centrally in your build system to avoid hundreds of build processes permanently pulling the git repo.

**Comment by clason** [comment]
Thank you for the explanation. 

> The last number in the semver is a monotonic timestamp, not a hash.

That's interesting, I failed to make that connection. I apologize if that is documented somewhere.

> DESTDIR must be an absolute path, use $PWD and not a dot.

Yes, I realized my mistake after posting this; of course the build works then.

**Pull request merged**